### PR TITLE
opt(gemm): add AITER MXFP4 preshuffle fast path

### PR DESCRIPTION
**Pull request by jasainio**
Reintroduces the GEMM-side preshuffle fast path that PR #335 dropped when consolidating FP4 GEMM dispatch. With AITER pinned and autotune disabled, `FP4GemmMXFunction` now emits already-shuffled scales / B-data from the fused quantize kernel and `gemm_fp4_impl(preshuffled=True)` calls `aiter.gemm_a4w4(bpreshuffle=True)` directly, skipping 3 per-GEMM shuffle kernel launches.

- ~10-15% per-GEMM speedup on Flux 12B shapes (microbench)
- Recovers the ~5% step-time regression reported on `flux_ddp_mxfp4_full_rebase_verify_263`
- End-to-end verified on Flux 12B MXFP4 1000-step (224 BpreShuffle kernel binds, exit 0) under `dev/mxfp4-aiter-preshuffle-min` @ 026e136
- HipBLASLt backend `can_handle` returns `False` on `preshuffled=True` so Path 1 raises cleanly instead of silently corrupting
- 4 commits: fast path + 3 supporting fixes (empty backend env var; `_clear_env_cache` re-export; FP4 GEMM `torch.compile` backward compat)

# Description

PR #335 consolidated FP4 GEMM dispatch but removed the preshuffle fast path, causing a ~5% step-time regression on Flux 12B MXFP4. This PR re-adds the fast path under a clean gate that activates only when (a) AITER is the pinned backend and (b) autotune is disabled -- the exact configuration where preshuffle is safe and beneficial.

When `enable_preshuffle()` returns `True`, the fused quantize kernel produces already-shuffled scales / B-data, and `gemm_fp4_impl(preshuffled=True)` dispatches directly to `aiter.gemm_a4w4(bpreshuffle=True)`. This skips 3 per-GEMM shuffle launches. HipBLASLt's `can_handle` returns `False` on `preshuffled=True` (the C++ extension has no preshuffle support), so any accidental routing to Path 1 raises a clean error instead of silently producing garbage.

The 3 supporting commits address: handling of empty `PRIMUS_TURBO_GEMM_BACKEND` env var, re-export of `_clear_env_cache` after the lru_cache removal, and FP4 GEMM backward compatibility under `torch.compile`.

Verified end-to-end as part of `dev/mxfp4-aiter-preshuffle-min` @ 026e136: Flux 12B MXFP4 1000-step exit 0 with 224 `BpreShuffle` AITER kernel binds confirmed.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring (performance optimization: re-introduces a fast path that PR #335 dropped)

## Changes

Please list the changes introduced in this PR:

- `primus_turbo/pytorch/kernels/gemm/gemm_fp4_impl.py`: add `enable_preshuffle()` helper (gates on AITER pin + autotune-off); thread `preshuffled: bool = False` through `custom_op`, `register_fake`, dispatcher kwargs, `make_key` (cache invalidation). AITER backend branches on `preshuffled` True/False; HipBLASLt backend `can_handle` returns `False` on `preshuffled=True`
- `primus_turbo/pytorch/ops/gemm_fp4.py`: wire `enable_preshuffle()` into the 6 `ScalingRecipe` constructors (only `shuffle_scale` + `shuffle_out` flip) and into the 3 `gemm_fp4_impl(...)` calls. `default_backend` switches to AITER under preshuffle. When deriving the col-wise transpose tensor and `shuffle_scale=True` is active, prefer the original BF16 input over `a_fp4.dequantize()` (which would consume the shuffled scale as canonical)
- `primus_turbo/pytorch/core/backend.py`: treat empty `PRIMUS_TURBO_GEMM_BACKEND` env var as missing; expose `_clear_env_cache` after the `lru_cache` removal
- `primus_turbo/pytorch/kernels/quantization/quantization_impl.py`: minor wiring for shuffled-output path
- Add 4 new tests in `tests/pytorch/ops/test_gemm_fp4.py`: `test_gemm_fp4_impl_aiter_preshuffle_parity` (bit-for-bit parity True vs False on Flux shapes), `test_gemm_fp4_impl_hipblaslt_preshuffle_opt_out` (locks HipBLASLt rejection contract), `test_gemm_fp4_mx_blockwise_quantized_tensor_aiter_preshuffled` (QuantizedTensorPair contract), and end-to-end wiring assertion in existing `test_gemm_fp4_mx_blockwise` matrix
- `tests/pytorch/core/test_global_backend_manager.py`: env-cache assertions
- `benchmark/ops/bench_gemm_fp4_preshuffle.py`: microbenchmark for cold/warm-cache preshuffle vs non-preshuffle

# Checklist:

- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (consumer docs updated in primus-tiger-training-internal `dev_workflow.md`)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes; end-to-end verified on Flux 12B MXFP4 1000-step (exit 0, 224 BpreShuffle kernel binds)

### CODE REVIEW

File: primus_turbo/pytorch/core/backend.py [hunk 1]
```diff
@@ -122,8 +124,20 @@ def _extract_backend_from_env(cls, env_value: str) -> Dict[PrecisionType, Backen
             for value in _PRECISION_TYPE_MAPPING.values():
                 precision_backend_dict[value] = BackendType[env_value.upper()]
 
+        cls._env_cache[env_value] = precision_backend_dict
         return precision_backend_dict
 
+    @classmethod
+    def _clear_env_cache(cls) -> None:
```
Comment (RuibinCheung) [inline primus_turbo/pytorch/core/backend.py:131]: Why need add this func ?

File: primus_turbo/pytorch/kernels/gemm/gemm_fp4_impl.py [hunk 2]
```diff
@@ -185,9 +207,39 @@ class GEMMFP4KernelDispatcher(AutoKernelDispatcher):
...
     @classmethod
-    def make_key(cls, a, b, trans_a, trans_b, trans_c, out_dtype, granularity, **kwargs):
+    def make_key(cls, a, b, trans_a, trans_b, trans_c, out_dtype, granularity, preshuffled=False, **kwargs):
         m, n, k = get_gemm_logical_shape(a, b, trans_a, trans_b)
-        return (m, n, k, a.dtype, b.dtype, out_dtype, trans_a, trans_b, trans_c, granularity)
+        return (m, n, k, a.dtype, b.dtype, out_dtype, trans_a, trans_b, trans_c, granularity, preshuffled)
+
+
+def enable_preshuffle() -> bool:
+    """Return True iff the AITER FP4 preshuffle fast path is safe to take.
```
Comment (RuibinCheung) [inline primus_turbo/pytorch/kernels/gemm/gemm_fp4_impl.py]: Could you try to simplify this comment ?

File: benchmark/ops/bench_gemm_fp4_preshuffle.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
+###############################################################################
```
Comment (RuibinCheung) [inline benchmark/ops/bench_gemm_fp4_preshuffle.py]: We want to keep the benchmark/ops directory simple. Could you removed this file?

**Review by RuibinCheung — CHANGES_REQUESTED** [review]

File: primus_turbo/pytorch/core/backend.py (hunk 1 above)
Comment (jasainio) [reply primus_turbo/pytorch/core/backend.py:131]: `_clear_env_cache()` replaces `@functools.lru_cache` which was removed because `lru_cache` on a classmethod is opaque to `torch._dynamo` and causes a graph break when `get_gemm_backend()` is called inside a `torch.compile` region. The call chain is: `enable_preshuffle()` → `GlobalBackendManager.get_gemm_backend()` → `_extract_backend_from_env()` — with `lru_cache` wrapping that last call, dynamo cannot trace through it and inserts a graph break.

The replacement is a plain `dict` class variable (`_env_cache`) that dynamo can trace through. The `_clear_env_cache()` method exposes the invalidation API so tests can reset state between parametrize runs (previously they called `_extract_backend_from_env.cache_clear()`).

If you prefer, I can inline the `.clear()` call directly into the test fixtures and drop the method.

File: primus_turbo/pytorch/kernels/gemm/gemm_fp4_impl.py (hunk 2 above)
Comment (jasainio) [reply primus_turbo/pytorch/kernels/gemm/gemm_fp4_impl.py]: Done — condensed to 5 lines in 291a7b0.

File: benchmark/ops/bench_gemm_fp4_preshuffle.py (hunk 3 above)
Comment (jasainio) [reply benchmark/ops/bench_gemm_fp4_preshuffle.py]: Done — removed in 291a7b0.

**Comment by wenchenvincent** [comment]
@jasainio  Does this PR target training or inference? For training, the data will need to be preshuffled every iteration.

**Comment by jasainio** [comment]
This targets **training**. The preshuffle is not a one-time preprocessing step -- it's fused into the quantize kernel that already runs every forward and backward call.

The existing `quantize_mxfp4` C++ kernel already accepts `shuffle_scale` / `shuffle_out` flags (added as part of the AITER MXFP4 integration before this PR). What this PR adds is the **wiring** that activates those flags:

- `FP4GemmMXFunction.forward` passes `shuffle_scale=True` / `shuffle_out=True` in the `ScalingRecipe` to `QuantizedTensor.quantize()` every iteration ([gemm_fp4.py L59-101](https://github.com/AMD-AGI/Primus-Turbo/blob/opt/gemm/fp4-aiter-preshuffle-fastpath/primus_turbo/pytorch/ops/gemm_fp4.py#L59-L101))
- The quantize kernel produces already-shuffled FP4 data + scales in a single fused pass
- `gemm_fp4_impl(preshuffled=True)` then calls `aiter.gemm_a4w4(bpreshuffle=True)` directly, skipping 3 standalone shuffle kernel launches that would otherwise run per-GEMM

The backward pass does the same: `grad_out` is re-quantized with `shuffle_scale=preshuffle` every iteration ([gemm_fp4.py L189-250](https://github.com/AMD-AGI/Primus-Turbo/blob/opt/gemm/fp4-aiter-preshuffle-fastpath/primus_turbo/pytorch/ops/gemm_fp4.py#L189-L250)).

So the cost model is: zero extra work (shuffle is folded into the quantize kernel we were already paying for), minus 3 kernel launches per GEMM.

File: primus_turbo/pytorch/core/backend.py (hunk 1 above)
Comment (RuibinCheung) [reply primus_turbo/pytorch/core/backend.py:131]: ACK

File: tests/pytorch/ops/test_gemm_fp4.py [hunk 4]
```diff
@@ -266,3 +298,418 @@ def test_gemm_fp4_mx_blockwise_quantized_tensor(m, n, k, layout, format, dtype,
...
+        # Both paths converge on the same aiter.gemm_a4w4(bpreshuffle=True)
+        # consuming the same byte-identical shuffled inputs -> outputs match.
+        assert torch.equal(out_no_pre, out_pre), (
+            "preshuffled=True must match preshuffled=False bit-for-bit " "on the AITER backend"
+        )
+    finally:
+        GlobalBackendManager.reset()
+
+
+def test_gemm_fp4_impl_hipblaslt_preshuffle_opt_out():
```
Comment (RuibinCheung) [inline tests/pytorch/ops/test_gemm_fp4.py]: Please remove this unit test case.

File: primus_turbo/pytorch/kernels/quantization/quantization_impl.py [hunk 5]
```diff
@@ -533,6 +533,9 @@ def quantize_mxfp4_impl(
     else:
         assert axis is None, "The axis must be None when with_trans is True."
 
+    if not x.is_contiguous():
```
Comment (RuibinCheung) [inline primus_turbo/pytorch/kernels/quantization/quantization_impl.py]: Dose x will be a non-contiguos tensor ? The contiguous may cause silent performance regression. If we need check x is contiguous please use assert `x.is_contiguous()`.

**Review by RuibinCheung — CHANGES_REQUESTED** [review]

File: tests/pytorch/ops/test_gemm_fp4.py (hunk 4 above)
Comment (jasainio) [reply tests/pytorch/ops/test_gemm_fp4.py]: Removed in bf138d3. The production guard in `GEMMFP4HipBLASLtBackend.supports()` that rejects `preshuffled=True` is untouched -- only the test is removed.

File: primus_turbo/pytorch/kernels/quantization/quantization_impl.py (hunk 5 above)
Comment (jasainio) [reply primus_turbo/pytorch/kernels/quantization/quantization_impl.py]: Good catch -- changed to `assert x.is_contiguous(), "The x tensor must be contiguous."` in bf138d3 to match the convention in the rest of the file (lines 51, 68, 454).

On `main`, `quantize_mxfp4_impl` has no contiguity check at all and the C++ kernel would silently produce wrong results on non-contiguous input. The assert makes the requirement explicit. The forward callers in `FP4GemmMXFunction` don't force contiguity on user inputs, but in practice Megatron/Primus-LM always feeds contiguous activations. If a non-contiguous case turns up, the fix would be adding `.contiguous()` at the caller level rather than hiding it inside the kernel.

**Review by RuibinCheung — DISMISSED** [review]
LGTM if CI is green.

**Review by RuibinCheung — APPROVED** [review]
LGTM
